dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/ToolDAO.java

```diff
@@ -51,4 +64,245 @@ public Tool findPublishedByToolPath(String path, String tool) {
         return uniqueResult(namedQuery("io.dockstore.webservice.core.Tool.findPublishedByToolPath").setParameter("path", path)
                 .setParameter("toolname", tool));
     }
+
+    /**
+     * Return map containing schema.org info retrieved from the specified tool's descriptor cwl
+     * @param id of specified tool
+     * @return map containing schema.org info to be used as json-ld data
+     */
+    public List findPublishedSchemaById(long id) {
+        Tool tool = findPublishedById(id);
+        List schemaObjects = new ArrayList();
+
+        if (tool != null) {
+            String defaultVersion = tool.getDefaultVersion();
+
+            String descriptorJson = getDescriptorJson(tool, defaultVersion);
+
+            String schemaVariable = getSchemaVariable(descriptorJson);
+
+            if (!schemaVariable.isEmpty()) {
+                Map<String, Object> map = new HashMap();
+                map = cwlJson2Map(descriptorJson, schemaVariable);
+
+                // must be done after json has been converted to json-ld
+                map = addContext(map, schemaVariable);
+                map = stripNamespace(map, schemaVariable);
+
+                schemaObjects = getSchemaObjectsList(map);
+            }
+        }
+
+        return schemaObjects;
+    }
+
+    /**
+     * Strip out properties to only get schema objects
+     * @param schemaMap properly annotated json-ld map with properties still attached
+     * @return list of schema objects without properties
+     */
+    private static List getSchemaObjectsList(final Map<String, Object> schemaMap) {
+        List schemaObjects = new ArrayList();
+        for (Map.Entry<String, Object> entry : schemaMap.entrySet()) {
+            Object value = entry.getValue();
+
+            if (value instanceof Map) {
+                schemaObjects.add(value);
+            } else if (value instanceof List) {
+                schemaObjects.addAll((List) value);
+            }
+        }
+        return schemaObjects;
+    }
+
+    /**
+     * Add @context to json-ld map and leaves out any schema properties which do not have a class
+     * @param schemaMap json-ld map
+     * @param schemaVariable string which refers to http://schema.org/
+     * @return a map that has been properly annotated with @context
+     */
+    private static Map<String, Object> addContext(final Map<String, Object> schemaMap, final String schemaVariable) {
+        Map<String, Object> merged = new HashMap<>(schemaMap);
+
+        for (Map.Entry<String, Object> entry : schemaMap.entrySet()) {
+            String key = entry.getKey();
+            Object copyValue = merged.get(key);
+
+            if (key.startsWith(schemaVariable)) {
+
+                if (copyValue instanceof List) {
+                    for (Iterator<Object> iterator = ((ArrayList) copyValue).iterator(); iterator.hasNext();) {
```

Comment:
This code doesn't seem to rely on ArrayList (as opposed to List).
Just use the abstract data type unless the specific type is required.